### DISCUSSION TITLE
Access a mesh variable from within a Function object

### DISCUSSION THREAD

**Discussion by bielsnohr** [idaholab/moose#17107] [Q&A General] [answered]
I am just starting to develop my own `MooseApp`, and I am trying to determine if it possible to access and return the value of a mesh variable at a given `Point` from within a `Function` object. The Doxygen documentation would seem to suggest not, but I am not yet very familiar with this expansive framework. I have seen an example where the mesh variable has been obtained by going through a `UserObject`, which is then accessed by the `Function`, but this strikes me as a bit roundabout.

For some context, my use case is to apply this `Function` within a `ADNeumannFunctionBC`. The mesh variable that the `Function` will pick up are heat flux values that will be coming from a subapp that calls an external program.

In case a diagram makes this a bit more clear (arrows are _loosely_ the flow of data I would like):

![Screenshot from 2021-02-23 17-23-25](https://user-images.githubusercontent.com/6177028/108882095-eb4f0e80-75fb-11eb-8be9-4286f68b7d61.png)

**Comment by aeslaughter** [comment, marked as answer]
What problem are you trying to solve? Are your main/sub-app meshes on the same scale? How is the "heat_flux" defined in the sub-application? 

Typically this type of value would be computed as an AuxVariable, if this is the case then you transfer the variable directly from the sub-app to the main using the MultiAppCopyTransfer (if your two problems are on the same mesh).

**Reply by bielsnohr** [reply]
> What problem are you trying to solve?

Standard transient heat conduction problem for a material (steel at this point). I'm using the HeatConduction module. The heat fluxes are coming from the fast ions emitted by a plasma. These are the values that will be returned by the external program.  

> Are your main/sub-app meshes on the same scale?

Spatial scale, yes. Time scale is still in the process of being determined. It is likely that the external program will run through a number of its own internal, smaller time steps and then return the heat flux at the end of the longer "global" time step for the MultiApp.

> How is the "heat_flux" defined in the sub-application?

Not sure of this because I have not actually had access to the external program yet. I appreciate that probably limits the advice you can give, but basically I have been trying to do as much on the MOOSE side of things while I wait for access.

> Typically this type of value would be computed as an AuxVariable, if this is the case then you transfer the variable directly from the sub-app to the main using the MultiAppCopyTransfer (if your two problems are on the same mesh).

It is quite likely the meshes will not be the same, but again, I need to see what the external program capabilities are and how it accepts a mesh. Presumably, I can design my MOOSE application such that it passes a mesh which is compatible with the external program, and therefore they will be on the same mesh. 

Once I have the AuxVariable, how can that be used in a Neumann BC?

Many thanks for the reply!

**Reply by bielsnohr** [reply]
> Once I have the AuxVariable, how can that be used in a Neumann BC?

Ah, looks like I might have found it: `CoupledVarNeumannBC` seems like it would do the trick.

**Reply by bielsnohr** [reply]
Although I will need to confirm meshing, this largely solves my use-case for the present time:

heat fluxes generated by external program -> MultiAppCopyTransfer -> AuxVariable -> CoupleVarNeumannBC

If the meshes are not identical for the Main and sub-app, then I believe all that will change is a custom Transfer will be needed.
